project-metrics/metrics_service/metrics/base.py

```diff
@@ -32,29 +40,81 @@ def label(self) -> Text:
   @property
   def formatted_result(self) -> Text:
     """The formatted result value (ex. 3w, 80%, 100PRs)."""
-    return self._format_value(self._result.value) if self._result else '?'
-
-  @property
-  def result(self) -> Optional[models.MetricResult]:
-    """The result of a computation of the metric."""
-    return self._result
+    return self._format_value(self.result.value) if self.result else '?'
 
   @property
   def score(self) -> models.MetricScore:
     """The 0-4 score for the metric."""
     return self._score_value(
-        self._result.value) if self._result else models.MetricScore.UNKNOWN
+        self.result.value) if self.result else models.MetricScore.UNKNOWN
+
+
+class Metric(MetricDisplay):
+  """Abstract base class for health metrics."""
+
+  __metaclass__ = abc.ABCMeta
+  __active_metrics = {}
+
+  @classmethod
+  def register(cls, metric_cls: Type[TypeVar('U', bound='Metric')]):
+    """Register a metric to be included in result fetching.
+
+    Args:
+      metric_cls: metric implementation to make active.
+    """
+    cls.__active_metrics[metric_cls.__name__] = metric_cls
+
+  @classmethod
+  def __from_result(cls, result: models.MetricResult):
+    """Wraps a result in its metric class for display/processing.
+
+    Expects that the result is for an active metric.
+
+    Raises:
+      KeyError: if the result is for an inactive or non-existent metric.
+
+    Args:
+      result: metric result retrieved from the DB.
+
+    Returns:
+      A metric implementation holding the result.
+    """
+    return cls.__active_metrics[result.name](result=result)
+
+  @classmethod
+  def get_latest(cls) -> Sequence['Metric']:
+    """Fetch the latest result for each metric."""
+    # TODO(rcebulko): Query DB.
+    metric_results = models.MetricResult.__table__
+    session = db_engine.get_session()
+
+    max_dates = session.query(
+        metric_results.c.name,
+        sqlalchemy.func.max(
+            metric_results.c.computed_at).label('max_computed_at')).group_by(
+                metric_results.c.name).filter(
+                    metric_results.c.name.in_(
+                        cls.__active_metrics.keys())).subquery('latest')
```

Comment:
1. Can you split `cls._active_metrics.keys` → `active_metrics_names` as temp variable for readability?
2. Some manual formatting here would go a long way towards readability - start every SQL-equivalent section in its own new line (so `.group_by`, `.filter` would start their own lines)

Reply:
1. Done.
2. Done. SQL Alchemy and Google-Format extension don't play very nice together :-1: